### ISSUE TITLE
why does query invoke a transaction

### ISSUE THREAD

**Issue by chenyuan53618** [sqlalchemy/sqlalchemy#5098]
mysqlclient==1.4.2.post1
SQLAlchemy==1.3.5

```
def test4():
    Session = db.session  # db.session = sessionmaker(bind=self.engine, **session_options)

    print("start")
    print(db.status())
    conn = db.engine.connect()
    print(db.status())
    print("after check out one connection")

    session = Session(bind=conn)
    new_user = User()
    new_user.id = 3
    print("begin the transaction")

    session.query(User).filter_by(id=1).first()


    # session.rollback() # uncomment this line. everything works fine
    print("invalidate the connection. mock the connection not available")
    conn.invalidate()

    print(db.status())
    session.query(User).filter_by(id=1).first()

    print(db.status())
```

this will generate exception:
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1177, in _execute_context
    conn = self._revalidate_connection()
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 463, in _revalidate_connection
    "Can't reconnect until invalid "
sqlalchemy.exc.InvalidRequestError: Can't reconnect until invalid transaction is rolled back
```
After this, this connection is no longer usable anymore.

My question is :
Why does query invoke a transaction?
Normally speaking, if there is no transaction, `conn.invalidate()` won't cause any problems. It just close the connection and return it back to pools, right? For example, if add `session.rollback()` right after query, everything works just fine.

**Comment by zzzeek** [comment]
the session does everything inside of a transaction that ends when you call commit(), rollback(), or close():  https://docs.sqlalchemy.org/en/13/orm/session_transaction.html

only if it's in "autocommit" mode will it use ad-hoc connections (which still means it's starting transactions, it just rolling them back immediately).   autocommit mode is going to be deprecated in 1.4 however.    there will be a new mode called "autobegin", which is basically what the default behavior is now.  however when autobegin is turned off, you will have to call session.begin() to start a transaction.  the difference from autocommit though it that it will disallow running queries without a transaction being started.

basically running SELECT statements in ad-hoc transactions that are auto-closed is an antipattern especially in the ORM that leads to bugs and confusion.

**Comment by jvanasco** [comment]
@zzzeek i'm very confused by this example, although I think it might how the exceptions are handled in `base.py` in this anti-pattern of doing things to the underlying connection 

the lines that get me is essentially this pattern, which is what I assume is causing the issue:

    session = Session(bind=conn)
    conn.invalidate()
    session.query(...

triggering this message:

    Can't reconnect until invalid transaction is rolled back

see: https://github.com/zzzeek/sqlalchemy/blob/master/lib/sqlalchemy/engine/base.py#L447-L459

Since the underlying connection to the Session was closed - shouldn't that have `closed` the "transaction"... OR shouldn't the Engine create a new connection then have the Session object raise an exception on detecting the connection changed mid-transaction/session?  This seems to bury the cause of the transaction being invalid, which could be so many reasons.

I'm sure there is a reason why things work this way, both under the hood and from an API perspective.

**Comment by zzzeek** [comment]
> @zzzeek i'm very confused by this example, although I think it might how the exceptions are handled in `base.py` in this anti-pattern of doing things to the underlying connection
> 
> the lines that get me is essentially this pattern, which is what I assume is causing the issue:
> 
> ```
> session = Session(bind=conn)
> conn.invalidate()
> session.query(...
> ```

the example above shows the query happening before the invalidate.  if you invalidated first, I think it would just transparently reconnect again because there was no transaction started yet.  the above example confirms this becuase when they rollback() first,then invalidate, they are invalidating the connection outside of a transaction, it then picks up on the next step without any issue.

> 
> triggering this message:
> 
> ```
> Can't reconnect until invalid transaction is rolled back
> ```
> 
> see: https://github.com/zzzeek/sqlalchemy/blob/master/lib/sqlalchemy/engine/base.py#L447-L459
> 
> Since the underlying connection to the Session was closed - shouldn't that have `closed` the "transaction"... 

invalidate is not the same as closed, it means, this connection had an unexpected problem.  if you were in the middle of a transaction you need to roll that back kind of as a sign that you aren't expecting the reconnected connection to pick up that same transaction, since it can't.    that is, it waits for you to confirm, hey, your transaction is gone, let us know you're in your "transaction failed" workflow before we start a new one.


> OR shouldn't the Engine create a new connection then have the Session object raise an exception on detecting the connection changed mid-transaction/session? 

the Connection is a fine place to note that an invalidate() was called after a transaction started and it was then not rolled back before subsequent use was attempted.  no reason the Session needs to get into that.

> This seems to bury the cause of the transaction being invalid, which could be so many reasons.

The cause of the transaction being invalid is that the sample code called connection.invalidate().   In real applications there are usually not use cases to call connection.invalidate() that are not inside of exception handlers already.   In fact there's really not a reason to call connection.invalidate() anyway as it is usually automatic by the "is_disconnect" handler, which again always occurs in the context of an exception throw.

i think the above example has the issue that it is not a real-world example, if they want to share what the *actual* use case is, we can talk about that too.

> 
> I'm sure there is a reason why things work this way, both under the hood and from an API perspective.

this is the pattern:

```
conn = engine.connect()
trans = conn.begin()
conn.execute("insert into table (some row)")
conn.invalidate()
conn.execute("insert into table (some other row)")  # <-- disallowed
trans.commit()

```

above, if we didn't raise on that second conn.execute(), what would the expectation be if not to raise a message, "hey, you just blew your transaction away via an esoteric method that typically indicates a failure mode, don't you want to be rolling back" ?       The only other thing that would work logically would be that the connection replayed all the INSERT statements leading up to that invalidate so that the transaction could be restored.   That's not a built-in behavior, I even wrote an extension for a client once who wanted that feature even though I advised against it.   

The best practice pattern for use of connection.invalidate(), if there were any, is this:

```

@retry(times=3)
def some_method():
    conn = engine.connect()
    trans = conn.begin()
    try:
        conn.execute("insert into table (some row)")
        conn.execute("insert into table (some other row)")
        trans.commit()
    except Exception as e:
       if this_exception_means_the_connection_is_dead(e):
            conn.invalidate()
       trans.rollback()
    finally:
        conn.close()

```

that is, if your transaction fails in the middle, and you want to recover, you need to roll up to the top and rerun the whole operation again.  the invalidate() is just a hint to the connection pool that this connection shouldn't be returned to the pool.     normally, is_disconnect() logic built into the engine does the work of "this_exception_means_the_connection_is_dead(e)" and the above invalidate is transparent.

**Comment by jvanasco** [comment]
thanks for the DETAILED response.  

I think I need to make an SSCCE example of this tonight to confirm my below thoughts.  

> In fact there's really not a reason to call connection.invalidate() anyway

100% in agreement. we're talking about an anti-pattern here...

I'm not concerned with allowing this to work, but how/when the exceptions are raised.

> invalidate is not the same as closed

The docs for `Connection.invalidate` state: (https://docs.sqlalchemy.org/en/13/core/connections.html#sqlalchemy.engine.Connection.invalidate) 

    The underlying DBAPI connection is literally closed (if possible), and is discarded. Its source connection pool will typically lazily create a new connection to replace it.

Based on that, I'm expecting the connection to most-likely be *implicitly* closed. 

However, I think this is the big confusion between us:

> if we didn't raise on that second conn.execute(), what would the expectation be if not to raise a message, "hey, you just blew your transaction away via an esoteric method that typically indicates a failure mode, don't you want to be rolling back" ?

I am 100% in agreement that we should be saying this.  The concern I was hoping to convey, is the raised Exception doesn't suggest the real cause of the error to me... which means I wouldn't know how to properly bugfix it.

When i read `Can't reconnect until invalid transaction is rolled back`, I am led to think about two potential "transactions":

* transaction as the unit of work in a SqlAlchemy Session
* transaction as begin/rollback/commit in the Sql

The given example, which we both agree is an anti-pattern for various reasons,  I'm a bit lost on how to interpret this particular exception message.  

If I interpret 'transaction' as a Sql concept,  why was there no implicit `close()` as the docs would suggest?

If I did a `rollback` + `begin`, wouldn't I want this to fail on the Session layer -- as I could have some "transaction" data within a single sqlalchemy Session unit-of-work coming from the first and second connections?  That doesn't seem like something I would want to allow.  

The `invalidate` example you offered makes perfect sense in the context sqlalchemy core, but when dealing with Session/ORM, it really seems like an invalidated connection should be fatal to a  Session which has any data in it (unless there was an explicit override), and the Session is what should be telling me there is an issue... not the Engine.

**Comment by zzzeek** [comment]
OK DMV-for-four-hours brain kicking in on that one, ill try to see if I can respond maybe on friday tomorrow I'm sort of out.

**Comment by jvanasco** [comment]
let me post a SSCCE before you respond

**Comment by jvanasco** [comment]
@zzzeek here's a gist with a SSCCE https://gist.github.com/jvanasco/39d7daad7cf8cbfa94102efbd7fb8636

I was initially worried about having dirty data and found a testcase to prove it... but then realized I missed putting the right `bind=conn` argument into the session.  unsurprisingly, SqlAlchemy does not let me put in bad data or mismanage data when I purposefully try to - the library has many defenses that work!

As the test cases in this sscce show though, the messaging in the raised exceptions doesn't necessarily reflect what is going on - or what the remedy is.

We're talking about invoking an anti-pattern here, where a session's underlying connection is closed or invalidated, and that is messing up an implicit/hidden transaction... so i'm not sure if this even needs to be addressed, but I think this is what tripped up the original poster.

If [`Connection.invalidate`](https://docs.sqlalchemy.org/en/13/core/connections.html#sqlalchemy.engine.Connection.invalidate) is invoked,  the exception raised is:

    sqlalchemy.exc.StatementError: (sqlalchemy.exc.InvalidRequestError) Can't reconnect until invalid transaction is rolled back

Comparing the docs (https://docs.sqlalchemy.org/en/13/core/connections.html#sqlalchemy.engine.Connection.invalidate) with the exception message, it don't think it is clear to an average user that we're dealing with use case from the 4th paragraph here ("If a transaction was in progress (e.g. the Connection.begin() method has been called) when Connection.invalidate() method is called...."), as this all happens behind the scenes and the `Transaction` object is unknown to the user.  I missed this on my first read.

if [`Connection.close`](https://docs.sqlalchemy.org/en/13/core/connections.html#sqlalchemy.engine.Connection.close) is called, the error generated makes perfect sense:

    sqlalchemy.exc.StatementError: (sqlalchemy.exc.ResourceClosedError) This Connection is closed
    
Since the original poster is using this in an ORM context, I'm not sure either of these exceptions are particularly helpful in debugging and solving the problem.  I think it may be better for the ORM to catch these exceptions, noting they must call [`Session.close`](https://docs.sqlalchemy.org/en/13/orm/session_api.html#sqlalchemy.orm.session.Session.close) or 
[`Session.invalidate`](https://docs.sqlalchemy.org/en/13/orm/session_api.html#sqlalchemy.orm.session.Session.invalidate) to recover.

**Comment by chenyuan53618** [comment]
> the session does everything inside of a transaction that ends when you call commit(), rollback(), or close(): https://docs.sqlalchemy.org/en/13/orm/session_transaction.html
> 
> only if it's in "autocommit" mode will it use ad-hoc connections (which still means it's starting transactions, it just rolling them back immediately). autocommit mode is going to be deprecated in 1.4 however. there will be a new mode called "autobegin", which is basically what the default behavior is now. however when autobegin is turned off, you will have to call session.begin() to start a transaction. the difference from autocommit though it that it will disallow running queries without a transaction being started.
> 
> basically running SELECT statements in ad-hoc transactions that are auto-closed is an antipattern especially in the ORM that leads to bugs and confusion.

**Comment by chenyuan53618** [comment]
Thx, @zzzeek. I just became your patron. Really worth all detailed answers

**Comment by zzzeek** [comment]
> The `invalidate` example you offered makes perfect sense in the context sqlalchemy core, but when dealing with Session/ORM, it really seems like an invalidated connection should be fatal to a Session which has any data in it (unless there was an explicit override), and the Session is what should be telling me there is an issue... not the Engine.

the session doesn't know anything about the state of the connection until it tries to use it for something.   then it gets that error.    so.....is this all, the session should wrap certain Core errors, try to interpret them and state them differently (that would require lots of new architecture to wrap all ORM SQL operations in a new try/catch container)?   are we just talking about an error message?  or perhaps that the Session should put itself into "DEACTIVE" state when it detects this?   all of these things would require major rearchitecting such that any SQL emitted anywhere on behalf of a Session would have to catch this, I'd likely use a handle_error event handler but this is all burden / complex / more testing overhead / performance overhead / more regressions / etc.      I'm not sure there's much of a problem here to warrant any of that.     if this were like a daily source of confusion for people, that would be one thing, but usually, if the database connection crashed, they get an error, and...their application either has retry logic built in, or it doesnt.  trying to work with the dead connection afterwards and being confused that the Session is not 100% blocking you does not seem like it would be worth lots of complexity.

**Comment by zzzeek** [comment]
closing this for now.
